h5io/tests/test_io.py

```diff
@@ -127,3 +127,72 @@ def test_numpy_values():
         assert_equal(read_hdf5(test_file, 'first'), value)
+
+
+def test_multi_dim_array():
+    trajectory = [[[[0.001, 0.01, 0.02, 0.03],
+                    [0.001, 0.11, 0.12, 0.13],
+                    [0.001, 0.21, 0.22, 0.23],
+                    [0.001, 0.31, 0.32, 0.33],
+                    [0.001, 0.41, 0.42, 0.43]],
+                   [[0.011, 0.01, 0.02, 0.03],
+                    [0.011, 0.11, 0.12, 0.13],
+                    [0.011, 0.21, 0.22, 0.23],
+                    [0.011, 0.31, 0.32, 0.33],
+                    [0.011, 0.41, 0.42, 0.43]],
+                   [[0.021, 0.01, 0.02, 0.03],
+                    [0.021, 0.11, 0.12, 0.13],
+                    [0.021, 0.21, 0.22, 0.23],
+                    [0.021, 0.31, 0.32, 0.33],
+                    [0.021, 0.41, 0.42, 0.43]]],
+                  [[[0.101, 0.01, 0.02, 0.03],
+                    [0.101, 0.11, 0.12, 0.13],
+                    [0.101, 0.21, 0.22, 0.23],
+                    [0.101, 0.31, 0.32, 0.33],
+                    [0.101, 0.41, 0.42, 0.43]],
+                   [[0.111, 0.01, 0.02, 0.03],
+                    [0.111, 0.11, 0.12, 0.13],
+                    [0.111, 0.21, 0.22, 0.23],
+                    [0.111, 0.31, 0.32, 0.33],
+                    [0.111, 0.41, 0.42, 0.43]],
+                   [[0.121, 0.01, 0.02, 0.03],
+                    [0.121, 0.11, 0.12, 0.13],
+                    [0.121, 0.21, 0.22, 0.23],
+                    [0.121, 0.31, 0.32, 0.33],
+                    [0.121, 0.41, 0.42, 0.43]],
+                   [[0.131, 0.01, 0.02, 0.03],
+                    [0.131, 0.11, 0.12, 0.13],
+                    [0.131, 0.21, 0.22, 0.23],
+                    [0.131, 0.31, 0.32, 0.33],
+                    [0.131, 0.41, 0.42, 0.43]],
+                   [[0.141, 0.01, 0.02, 0.03],
+                    [0.141, 0.11, 0.12, 0.13],
+                    [0.141, 0.21, 0.22, 0.23],
+                    [0.141, 0.31, 0.32, 0.33],
+                    [0.141, 0.41, 0.42, 0.43]]],
+                  [[[0.201, 0.01, 0.02, 0.03],
+                    [0.201, 0.11, 0.12, 0.13],
+                    [0.201, 0.21, 0.22, 0.23],
+                    [0.201, 0.31, 0.32, 0.33],
+                    [0.201, 0.41, 0.42, 0.43]],
+                   [[0.211, 0.01, 0.02, 0.03],
+                    [0.211, 0.11, 0.12, 0.13],
+                    [0.211, 0.21, 0.22, 0.23],
+                    [0.211, 0.31, 0.32, 0.33],
+                    [0.211, 0.41, 0.42, 0.43]],
+                   [[0.221, 0.01, 0.02, 0.03],
+                    [0.221, 0.11, 0.12, 0.13],
+                    [0.221, 0.21, 0.22, 0.23],
+                    [0.221, 0.31, 0.32, 0.33],
+                    [0.221, 0.41, 0.42, 0.43]],
+                   [[0.231, 0.01, 0.02, 0.03],
+                    [0.231, 0.11, 0.12, 0.13],
+                    [0.231, 0.21, 0.22, 0.23],
+                    [0.231, 0.31, 0.32, 0.33],
+                    [0.231, 0.41, 0.42, 0.43]]]]
```

Comment:
It is hard to immediately see what this is supposed to create. How about doing something similar:
```
rng = np.random.RandomState(0)
trajectory = [rng.randn(2,1), rng.randn(3, 1)]
```
... or whatever is the simplest thing that shows what works.

Then also please add tests for ones that should fail, which I guess is what here should hit the error in `_validate_sub_shapes`.

Also (this might be the same thing?) but this code should gracefully handle trying to save `bad = np.array([np])` and similar non-savable variants.